docs/standards/secure.md

```diff
@@ -0,0 +1,49 @@
+# Secure Coding Standards
+
+## Introduction
+
+CiviCRM maintains a number of standard practices which help ensure that CiviCRM is as secure as possible. This chapter will aim to help give developers guidance on the best way to write code for CiviCRM core and Extensions etc in a secure way. 
+
+## Escape on Input v Escape on Output.
+
+Escaping on input means that developers ensure that every single input from their Interface(s) are properly escaped before passing them into the database. This has a major issue for an application like CiviCRM because there are too many various interfaces to try and do proper escape on Input. There is also a risk that when you escape on input you can dramatically change the value and strip out some data through the escaping process.  Where as escaping on output means you have to cover all your various interfaces, ensure that all of them properly and safely account for the possibility that there maybe unsafe data in your database and sanitise it for safe viewing / usage in for example HTML or AngularJS templating. 
+
+CiviCRM has long been confused and staggered in regards to whether to escape on output or escape on input. CiviCRM are slowly moving towards escaping on output for most purposes however there is still a need for escaping on input when dealing with writing queries against the database. At present the simplest way to escape on output is to use inbuilt escape functions within our templating engine Smarty. For example:
+
+```
+<a  href="{$item.url}" title="{$item.title|escape:'html'}">
+```
+
+This will ensure that the variable title within the item key when generating a list of recently viewed items won't have any Cross Site Scripting as it will be escaped for use within HTML. For more information on the types of escaping you can do with Smarty see the [Smarty Documentation](https://www.smarty.net/docsv2/en/language.modifier.escape)
+
+However sometimes to escape on output you need to ensure that because of the complex nature of the variable that the variable is properly escaped when passed to Smarty. For example, when building a json encoded blob of data for use in an contribution form it was necessary to escape before passing onto the Smarty Template.
+
+```php
+$form->assign('submittedOnBehalfInfo', json_encode(str_replace('"', '\"', $form->_submitValues['onbehalf']), JSON_HEX_APOS));
+```
```

Comment:
@seamuslee001 I'm a little confused about this code block and worried that it might be a bad example. But maybe I'm missing something. Here's my thinking: 

* My understanding is that `str_replace('"', '\"', $untrustedData)` is bad practice because an attacker can pass a string like `foo\" bar` and effectively "escape the escape". So reading that threw up a red flag for me.
* But then I see the `json_encode` function and think *"oh, ok, that's good"*. I don't really understand why `str_relace` is there. Maybe we could use a simpler example here? Or explain more?